### ISSUE TITLE
100% CPU usage running greeter example?

### ISSUE THREAD

**Issue by nathanielmanistaatgoogle** [grpc/grpc#9688]
[A user writes](https://groups.google.com/forum/#!topic/grpc-io/hzn6eKDSMlM):
```
I was trying out the greeter example in the repository and I
noticed that the server uses 100% CPU after the first request.
This only happens when started with Python 2, and not with
Python 3.

I'm using Python 2.7.13 on Arch Linux, grpc is installed with
pip in a fresh virtualenv. Does anyone experience the same
problem?
```

**Comment by alexf91** [comment]
Hi, thanks for the response.

I'm using grpcio 1.1.0.
I can reproduce the problem every time by starting greeter_server.py and then greeter_client.py. The client receives the response and terminates, and then the CPU load of the server jumps to 100%. It still responds to requests after that.

Profiling results with yappi:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   28.673   28.673 threading.py:743(CleanupThread.run)
        1    0.000    0.000   28.673   28.673 _server.py:662(_serve)
        1    0.001    0.001    0.906    0.906 greeter_server.py:30(<module>)
        5    0.004    0.001    0.821    0.164 __init__.py:29(<module>)
        1    0.004    0.004    0.811    0.811 __init__.py:16(<module>)
        2    0.000    0.000    0.596    0.298 __init__.py:3019(_call_aside)
        1    0.000    0.000    0.596    0.596 __init__.py:3036(_initialize_master_working_set)
       40    0.000    0.000    0.591    0.015 __init__.py:682(WorkingSet.add_entry)
       64    0.007    0.000    0.588    0.009 __init__.py:2003(find_on_path)
       30    0.000    0.000    0.563    0.019 __init__.py:1977(_by_version_descending)
```
Running the server with `time` for 30 seconds shows that most of the time is spent in kernel space:
```9.79s user 18.60s system 93% cpu 30.403 total```

It looks like it is spinning on a lock or something while waiting for new events.

**Comment by nathanielmanistaatgoogle** [comment]
I am able to reproduce this problem with Python 2.7. @mehrdada have you yet taken a look?

**Comment by mehrdada** [comment]
I did a quick test on macOS/Python2.7 and was unable to observe the behavior. Need to investigate further.

**Comment by mehrdada** [comment]
I looked further into the issue and it seems it is fixed in the latest master (86e9f3564a9ab05ac14d81628167bc359ca56694). @alexf91 can you please test to see if you can observe the issue on master?

**Comment by alexf91** [comment]
Yes, master is working fine. Only checked with Python 2.7 though.

**Comment by nathanielmanistaatgoogle** [comment]
@alexf91: thank you for looking. How severely are you affected by this problem?

We believe that a release will be cut from (what today is) `master` in less than six weeks.

**Comment by alexf91** [comment]
I'm not affected at all. I just started using grpc and can use my own build for development.

**Comment by mehrdada** [comment]
I tried Ubuntu 16.04/Python 3.5 as well and the issue does not seem to affect master.

**Comment by nathanielmanistaatgoogle** [comment]
Given what's been said so far, I think the right thing for this bug is to leave it until the 1.2.0 release and then close it. While it's definitely weird we're not currently aware of it being a large problem for anyone. We're curious but finding the root cause would be a big investment into [somatic](https://en.wikipedia.org/wiki/Somatic_(biology)) code.

**Comment by scottbez1** [comment]
I believe I've run into this issue when attempting to port a small system to gRPC, which is going to prevent me from making further progress on that. Any other info on an ETA for a patched stable release?

**Comment by nicolasnoble** [comment]
1.2.0 ought to be released in about a month now.

**Comment by kernelpig** [comment]
the same problem: run in docker with centos, python 2.7, pip installed grpcio(1.1.3), not found in grpcio-v0.15. why must use thread pool in grpcio(1.1.3)? why not make it as optional param....  sorry for Chinglish...

**Comment by nathanielmanistaatgoogle** [comment]
@kernelpig: the thread pool was made a required, positional parameter of server construction because we believed that we always wanted applications to have that level of control over how servers serviced concurrent RPCs.

**Comment by spielmannj** [comment]
I can reproduce the behaviour with grpcio 1.1.5 on Python **3.6.0** (compiled locally) on Ubuntu 16.04.

Master is not affected.

**Comment by sofo23** [comment]
I have the same problem

**Comment by kernelpig** [comment]
@nathanielmanistaatgoogle thank you for your answer

**Comment by kernelpig** [comment]
@sofo23 master has solved this problem. I tried it in python 2.7, it's ok.

**Comment by sofo23** [comment]
@kernelpig Thank you for your reply.

 I try to install using this command grpcio using this command:
pip install git+git://github.com/grpc/grpc.git@86e9f3564a9ab05ac14d81628167bc359ca56694
Is it the right way to install the master branch? It takes soo much time.

**Comment by egoldschmidt** [comment]
Weirdly, we are seeing this issue on Ubuntu + Python 3.5.2 up to grpc 1.2.0rc2 but 86e9f35 appears to be working great.

**Comment by nathanielmanistaatgoogle** [comment]
@egoldschmidt: if it's happening for you in a reproducible and deterministic way with 1.2._anything_ we'd love to hear about it in as much detail as you can share.

**Comment by egoldschmidt** [comment]
Call me crazy but I think I have a repro case here: https://github.com/egoldschmidt/go-build-time-test

I threw it together quickly so apologies if I'm missing something!

**Comment by assiotis** [comment]
I'm configmring that I can repro the problem consistently with v1.1.0, v1.1.3, v1.2rc1, v1.2rc2 all installed through pip. I'm running Ubuntu 16.04 and python 2.7.12. As soon as one request completes, a single CPU core maxes out to 100% and stays there until I manually restart the server.

**Comment by dgn** [comment]
we can still reproduce this with latest grpc release v1.2.0 - sadly keeps us from using grpc in production

**Comment by alexf91** [comment]
Is there something wrong with the wheels on pypi?

I've built the tagged 1.1.0 and 1.2.0 versions, and both work fine. Then I built it from the 1.2.0 source archive on pypi, which also works fine.
The load only goes up when I use the wheels from pypi.

**Comment by timmyzaid** [comment]
I've been experimenting with this and might have a stack trace that could be helpful. 

`#0  0x00007f384ffdfaed in poll () at ../sysdeps/unix/syscall-template.S:81
#1  0x00007f384c978526 in pollset_work (exec_ctx=0x7f3849956380, pollset=0x1d48cc0,
    worker_hdl=0x7f3849956310, now=..., deadline=...) at src/core/lib/iomgr/ev_poll_posix.c:984
#2  0x00007f384c995369 in grpc_completion_queue_next (cc=0x1d48bd0, deadline=...,
    reserved=reserved@entry=0x0) at src/core/lib/surface/completion_queue.c:429
#3  0x00007f384c93bbe1 in __pyx_pf_4grpc_7_cython_6cygrpc_15CompletionQueue_2poll (
    __pyx_v_deadline=<optimized out>, __pyx_v_self=0x7f384fd630d0)
    at src/python/grpcio/grpc/_cython/cygrpc.c:10417
#4  __pyx_pw_4grpc_7_cython_6cygrpc_15CompletionQueue_3poll (__pyx_v_self=0x7f384fd630d0,
    __pyx_args=<optimized out>, __pyx_kwds=<optimized out>)
    at src/python/grpcio/grpc/_cython/cygrpc.c:10277
#5  0x00007f3850c9ae8e in _PyCFunction_FastCallDict (func_obj=0x7f384b3edb40,
    args=0x7f3844001258, nargs=<optimized out>, kwargs=kwargs@entry=0x0)
    at Objects/methodobject.c:231`

I think that python executable and grpc executable are waiting on the same semaphore to unlock and it's causing them both to just spin, deadlocked.

`  3    Thread 0x7fded4e59700 (LWP 166) "python" 0x00007fdedb13eaed in poll () at ../sysdeps/unix/syscall-template.S:81
  2    Thread 0x7fdecffff700 (LWP 173) "python" sem_wait () at ../nptl/sysdeps/unix/sysv/linux/x86_64/sem_wait.S:85
* 1    Thread 0x7fdedc47a700 (LWP 164) "python" 0x00007fdedb140893 in select () at ../sysdeps/unix/syscall-template.S:81`

**Comment by timmyzaid** [comment]
Also the packages that are built when you pip install from the source versus when you pip install from pypi are binary different. I can give you a list of files that are different between the two packages if that would help.

**Comment by timmyzaid** [comment]
When I install the pypi package, src/core/lib/iomgr/ev_poll_posix.c is used by src/core/lib/surface/completion_queue.c, however, when I build src/core/lib/iomgr/ev_poll_linux.c is used. Not sure why or what the difference is yet.

**Comment by nathanielmanistaatgoogle** [comment]
@dgn: on what platform are you working?

@egoldschmidt, @assiotis, @alexf91, @timmyzaid: thank you for digging so deeply into this!

**Comment by alexf91** [comment]
To sum up my results:
manylinux builds have the issue, others don't (at least not on my machine). I've built the current master in the manylinux Docker container, and it also has the issue.

The reason why I did not see this issue with Python 3.6 when I reported this, is because there was no manylinux build for 1.1.0 and Python 3.6, so this was built from the source archive.

**Comment by timmyzaid** [comment]
I'm not sure what your build system looks like but I wonder if `GRPC_POLL_STRATEGY` is set on it which is causing the event engine to always be the posix version? Outside of that, I'm pretty sure a debug build can be built with that set to `poll` to be able to reproduce it from source.

**Comment by nathanielmanistaatgoogle** [comment]
We (mostly @nicolasnoble) think we have a diagnosis and I'd like us to have a fix today-tomorrow-ish. We'll definitely be pushing a patch release once the fix is implemented and integrated.

Again, thank you for your investigations.

**Comment by nicolasnoble** [comment]
Well, we'll have to make sure we fully understand the problem first :-) The diagnosis may sound good, but pinpointing the exact problem and coming with a resolution might be harder.

**Comment by dgn** [comment]
@nathanielmanistaatgoogle: we installed using the 1.2.0 wheels from PyPI on Centos 7

If I can find the time to build from source, I'll give an update on my findings

**Comment by ganesh-karthick** [comment]
@nathanielmanistaatgoogle  I am seeing this issue on Centos 7 Python 2.7.11 , would it be possible share what was the diagnosis of the issue?

**Comment by comzyh** [comment]
same issue on Ubuntu 16.04/Python 3.5.2/install from pip


I run `strace` on the grpc process and get output like that:

```
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=5, events=POLLIN|POLLOUT}, {fd=11, events=0}], 4, 188) = 1 ([{fd=11, revents=POLLHUP}])
clock_gettime(CLOCK_MONOTONIC, {905930, 24155064}) = 0
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=5, events=POLLIN|POLLOUT}, {fd=11, events=0}], 4, 188) = 1 ([{fd=11, revents=POLLHUP}])
clock_gettime(CLOCK_MONOTONIC, {905930, 24206581}) = 0
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=5, events=POLLIN|POLLOUT}, {fd=11, events=0}], 4, 188) = 1 ([{fd=11, revents=POLLHUP}])
....
```
It seems that after a TCP connection is closed, `poll` will return `POLLHUP` on that fd. 
But grpc will keep `poll`ing this fd(closed connection), and always get `POLLHUP` immediately after `poll`, and so enter infiniteloop.

**Comment by kpayson64** [comment]
@comzyh Your analysis was correct, we were not removing fds that were shutdown from the pollset.

I've created a fix in #10345 and pushed a 1.2.1 release.

**Comment by comzyh** [comment]
I tried 1.2.1 release, it did solved the this issue on a simple echo server. 

Cheers.

but when I try 1.2.1 on our production service, CPU 100% happened again.

I run `strace` again on the thread that cost 100% CPU (by `top -H -p <pid>` ), the output seems like
```
poll([{fd=3, events=POLLIN}, {fd=8, events=POLLIN}, {fd=5, events=POLLIN|POLLOUT}, {fd=70, events=0}, {fd=71, events=0}, {fd=73, events=0}, {fd=85, events=0}, {fd=87, events=0}, {fd=88, events=0}, {fd=89, events=0}, {fd=90, events=0}, {fd=91, events=0}, {fd=92, events=0}, {fd=93, events=0}, {fd=94, events=0}, {fd=95, events=0}, {fd=96, events=0}, {fd=97, events=0}, {fd=98, events=0}, {fd=99, events=0}, {fd=100, events=0}, {fd=101, events=0}, {fd=102, events=0}, {fd=103, events=0}, {fd=104, events=0}, {fd=105, events=0}, {fd=106, events=0}, {fd=107, events=0}, {fd=108, events=0}, {fd=109, events=0}, {fd=110, events=0}, {fd=111, events=0}, ...], 775, 71) = 772 ([{fd=70, revents=POLLHUP}, {fd=71, revents=POLLHUP}, {fd=73, revents=POLLHUP}, {fd=85, revents=POLLHUP}, {fd=87, revents=POLLHUP}, {fd=88, revents=POLLHUP}, {fd=89, revents=POLLERR|POLLHUP}, {fd=90, revents=POLLHUP}, {fd=91, revents=POLLHUP}, {fd=92, revents=POLLHUP}, {fd=93, revents=POLLHUP}, {fd=94, revents=POLLHUP}, {fd=95, revents=POLLHUP}, {fd=96, revents=POLLHUP}, {fd=97, revents=POLLHUP}, {fd=98, revents=POLLHUP}, {fd=99, revents=POLLHUP}, {fd=100, revents=POLLHUP}, {fd=101, revents=POLLHUP}, {fd=102, revents=POLLHUP}, {fd=103, revents=POLLHUP}, {fd=104, revents=POLLERR|POLLHUP}, {fd=105, revents=POLLHUP}, {fd=106, revents=POLLHUP}, {fd=107, revents=POLLHUP}, {fd=108, revents=POLLHUP}, {fd=109, revents=POLLHUP}, {fd=110, revents=POLLHUP}, {fd=111, revents=POLLHUP}, ...])
```
I got `POLLHUP` issue again.

by the way, as If I **force compile** with `--no-binary` when installing grpc, I won't encounter this issue.

> pip3 install grpcio==1.2.1 --ignore-installed --no-binary grpcio

As many @spielmannj   mentioned above, install from source code or install with compiling (even on earlier version) can always avoid this issue.

I'm trying further test to make sure whether it's my own fault, or #10345 doesn't totally fix this bug.

Just let you know.

**Comment by timmyzaid** [comment]
@comzyh as I mentioned earlier when you build and install locally `GRPC_POLL_STRATEGY` isn't set and so a different polling engine is used than when you install from pip. It still seems like that environment variable is set on the build servers which forces it to always use the posix polling engine.

Additionally, @kpayson64, I'm also still seeing the issue on some of my calls but not others. Trying to narrow down the difference.

**Comment by timmyzaid** [comment]
@nathanielmanistaatgoogle any updates? I'm pretty sure this should be re-opened.

**Comment by nathanielmanistaatgoogle** [comment]
Reopening; @kpayson64 is actively working on this issue.

The problem has turned out to be more subtle than we first anticipated and the scope of fix we're looking at making is looking more appropriate to a minor version release than a point release, so right now we're aspiring for this to be fixed in 1.3.0 (mid-to-late April).

**Comment by comzyh** [comment]
@timburks yes, days ago, I tried to attach grpc thread with GDB, and found that it's calling `poll` from `ev_poll_posix.c` rather than `ev_epoll_linux.c`. I know why now, thank you.

**Comment by timmyzaid** [comment]
Any updates here?

**Comment by Duiesel** [comment]
Hello!
Same problem to me.
Any updates?
Thanks!

**Comment by kpayson64** [comment]
For people still experiencing this issue post 1.2.1, are you able to provide a simple reproduction case?

**Comment by kevintsai** [comment]
same issue on CentOS 6.5 / Python 3.5.2 / grpcio 1.3.0 installed by pip3.
The result of analysis as same as @comzyh did.

Reproduction:
One of multiple clients close the connection and they're calling a response-streaming API like:
```
rpc MyLongStreamingAPI(UserInfo) returns(stream UserStatus) {}
```

**Comment by mengtint** [comment]
still seeing this on ubuntu16, grpcio==1.3.0
it happens unpredictably.
I have multiple golang grpc client create a new connection to python grpc server and close connection when job is done.

**Comment by timmyzaid** [comment]
You guys realize we're all pinned at 1.0.4 until this is fixed? And that it makes recommending grpc extremely difficult? The last real update was that you guys had a strategy for fixing it, does that still hold? I've put off trying to look at it more myself since it was implied that you were close to a fix.

**Comment by timmyzaid** [comment]
When there is a call that creates the issue, poll ends up spinning on a locked socket that isn't getting destroyed in chttp2_transport.c::destructive_reclaimer_locked because it still has references. grpc_chttp2_ref_transport and grpc_chttp2_unref_transport are called the same number of times for good and bad calls so there is something else that is messing with the reference count, trying to track that down.

**Comment by timmyzaid** [comment]
There's some sort of reference cycle, forcing the garbage collector to collect using collect() after event = None fixes the problem.

**Comment by timmyzaid** [comment]
The code as is relies on python cleaning up the objects. In the case of a reference cycle, the objects won't be deleted until the garbage collector runs, which is non-deterministic. To insure proper destruction of objects, the python logic needs to be restructured to explicitly clean up objects instead of relying on cascading deletes. https://stackoverflow.com/questions/15974561/i-cant-get-dealloc-be-called-when-deleting-an-object sheds some light on what I'm thinking causes the problem.

**Comment by timmyzaid** [comment]
Of course you could try to figure out what is causing the reference cycle as the GC is only needed when there are cycles, but there is no obvious one and there are a number of references so it would be difficult to track down. 

As an aside, having event = None really just masks the real problem. It amounts to tricking python into deleting the object when the real problem is that proper clean up of request_call isn't happening, resulting in the port hanging open.

**Comment by ofpiyush** [comment]
Can confirm seeing this with 1.4.0 release.

Also happened with the solution @comzyh proposed.

```
pip install grpcio==1.4.0 --ignore-installed --no-binary grpcio
```

We'll have to put resource limits on these processes.

Bleeding close to 25 usd a month on every gRPC process isn't fun...

**Comment by egoldschmidt** [comment]
Still seeing this. Any update on status? I'm kind of shocked such a serious issue has been outstanding for so long.

**Comment by timmyzaid** [comment]
Ruby got a fix for a similar sounding issue... https://github.com/grpc/grpc/pull/11764

**Comment by criver** [comment]
We are seeing 100%CPU issue on `1.3.0`. Any workaround on this except for reverting back to `1.0.4`?

**Comment by chenditc** [comment]
@criver I was able to workaround this by using:

```
pip3 install grpcio --ignore-installed --no-binary grpcio
```

**Comment by criver** [comment]
@chenditc The workaround works for us. Thanks.

**Comment by conanchen** [comment]
The same problem on GRPC-JAVA V1.6.1 on MacOS. I am debugging my IDEA plugin , the plugin will retrieve images' meta data  and then access the image with http.  below list the resouces usages under several scenarios :- 

CPU             Memory        Threads          GRPC         HTTP   
0.3 ~2.7       658M               39                   W/O           W/O
2.2 ~ 12.8    659M               40                   W/O           W/
74.5 ~ 279   1.55G              50~60              W/             W/

If I turn on GRPC feature, Intellij IDEA slow down dramatically, GRPC consumes too much CPU & RAM.

**Comment by kpayson64** [comment]
I've attempted an alternate fix and released a 1.6.3 patch release.  If this is still an issue, please reopen.

**Comment by ccl0326** [comment]
We have update grpc to 1.6.3. But i think some problems still exists.

We have 3 grpc instances.

The perf for bad one.
```
  71.40%  [kernel]              [k] _raw_spin_lock
   3.71%  [kernel]              [k] _raw_spin_unlock_irqrestore
   1.29%  [kernel]              [k] sk_run_filter
   0.67%  python2.7             [.] 0x000000000005735d
   0.42%  python2.7             [.] 0x0000000000082b0d
   0.37%  python2.7             [.] 0x000000000005f1f0
   0.31%  [kernel]              [k] finish_task_switch
   0.26%  python2.7             [.] 0x000000000012de6a
   0.22%  python2.7             [.] 0x00000000000b0f49
   0.22%  python2.7             [.] 0x000000000006009a
   0.20%  python2.7             [.] 0x00000000000b310d
   0.20%  libpthread-2.17.so    [.] __GI___pthread_cleanup_upto
   0.18%  python2.7             [.] 0x000000000006014c
   0.17%  python2.7             [.] 0x00000000000b1f80
   0.16%  [kernel]              [k] __do_softirq
   0.16%  python2.7             [.] 0x00000000000b272c
   0.15%  libc-2.17.so          [.] realloc_check
   0.14%  python2.7             [.] 0x000000000005a7df
   0.14%  python2.7             [.] 0x00000000000b2700
   0.13%  libc-2.17.so          [.] _int_free
   0.12%  libc-2.17.so          [.] free_check
```

The perf for normal one.
```
   9.92%  [kernel]             [k] _raw_spin_unlock_irqrestore
   3.46%  [kernel]             [k] sk_run_filter
   2.56%  python2.7            [.] 0x000000000005735d
   1.57%  python2.7            [.] 0x0000000000082b0d
   1.20%  python2.7            [.] 0x000000000005f1f0
   1.01%  python2.7            [.] 0x000000000012de6a
   1.00%  python2.7            [.] 0x00000000000b0f49
   0.98%  python2.7            [.] 0x000000000006009a
   0.90%  [kernel]             [k] finish_task_switch
   0.88%  libc-2.17.so         [.] _int_malloc
   0.79%  libc-2.17.so         [.] malloc_atfork
   0.72%  python2.7            [.] 0x00000000000b310d
   0.70%  libpthread-2.17.so   [.] do_futex_wait
   0.64%  libpthread-2.17.so   [.] _pthread_cleanup_pop_restore
   0.63%  python2.7            [.] 0x00000000000b272c
   0.63%  python2.7            [.] 0x00000000000b1f80
   0.62%  python2.7            [.] 0x000000000006014c
   0.59%  python2.7            [.] 0x00000000000b2700
   0.58%  python2.7            [.] 0x000000000005a7df
   0.58%  [kernel]             [k] iowrite16
   0.55%  python2.7            [.] 0x0000000000079305
```

And the traffic is balanced. I think maybe some data race problem here in grpc c-core?

**Comment by ccl0326** [comment]
ping @kpayson64

**Comment by kpayson64** [comment]
@ccl0326 

There isn't enough information here to conclude that this is the same as the issue that was initially reported.  Does the CPU reach 100% on the bad instance even when you stop serving traffic from it?

**Comment by ccl0326** [comment]
@kpayson64 
You'r right. I will open a new issue for it.

**Comment by achidlow** [comment]
I'm still seeing this on 1.7.0. Bad thread is using 100% CPU even after stopping serving traffic. Running `strace` on that particular thread shows the same output as comzyh was getting [here](https://github.com/grpc/grpc/issues/9688#issuecomment-289194383).

**Comment by achidlow** [comment]
@kpayson64 this is definitely still happening in 1.7.0 ^

**Comment by mdsf22** [comment]
I'm still seeing this on 1.8.0, centos7.3  python 2.7.13

**Comment by kpayson64** [comment]
I'm going to try another fix for the upcoming 1.8 release happening later this month.

**Comment by mehrdada** [comment]
Reopening because the fix causes macOS test failures and has been reverted on master

**Comment by kpayson64** [comment]
Rolled forward in #13665
